docs/src/modules/ROOT/pages/constraints-and-score/score-calculation.adoc

```diff
@@ -1622,6 +1622,82 @@ This stream now contains all `Employee` instances, both with and without shifts
 and can be passed to the <<collectorsLoadBalance,load balancing constraint collector>>,
 which will compute the unfairness metric.
 
+[#constraintStreamsPrecompute]
+=== Precomputing a stream
+
+When a stream depends only on properties of problem facts or entities, that stream can be precomputed to improve performance.
+Consider the following "no overlapping shifts" constraint:
+
+[tabs]
+====
+Java::
++
+[source,java,options="nowrap"]
+----
+Constraint noOverlappingShifts(ConstraintFactory constraintFactory) {
+    return constraintFactory.forEachUniquePair(Shift.class, Joiners.overlapping(Shift::getStart, Shift::getEnd), Joiners.equal(Shift::getEmployee))
+        .penalize(HardSoftScore.ONE_HARD, (shift1, shift2) -> shift1.getOverlap(shift2))
+        .asConstraint("No overlapping shifts");
+}
+----
+====
+
+If Shift's `start` and `end` are not variables (genuine or shadow), and Shift's `getOverlap` does not depend on any variables, then some parts of the constraint can be precomputed to improve performance:
+
+[tabs]
+====
+Java::
++
+[source,java,options="nowrap"]
+----
+TriConstraintStream<Shift, Shift, Integer> overlappingShifts(PrecomputeFactory precomputeFactory) {
+    return precomputeFactory.forEachUnfilteredUniquePair(Shift.class, Joiners.overlapping(Shift::getStart, Shift::getEnd))
+        .expand((shift1, shift2) -> shift1.getOverlap(shift2));
+}
+
+Constraint noOverlappingShifts(ConstraintFactory constraintFactory) {
+    return constraintFactory.precompute(this::overlappingShifts)
+        .filter((shift1, shift2, overlap) -> shift1.getEmployee() != null && shift1.getEmployee() == shift2.getEmployee())
+        .penalize(HardSoftScore.ONE_HARD, (shift1, shift2, overlap) -> overlap)
+        .asConstraint("No overlapping shifts");
+}
+----
+====
+
+[IMPORTANT]
```

Comment:
I don't think this deserves an `IMPORTANT` - it's just a normal description of the feature. The next bit - the score corruption - that is IMO correctly important; it's a typical caveat that people will run into.

Don't overdo it with highlighting - if everything is highlighted, nothing is.